File: harness/semgrep/fixtures/failopen/store.bad.ts
// BAD: replay/session store defaults to an in-memory impl when the configured
// shared backing is absent, or is hard-wired with no injectable alternative.

import { createMemorySessionStore } from './store';

interface EngineConfig {
    store?: SessionStore;
}

export function createEngineNullish(config: EngineConfig) {
    // nullish default => multi-instance fails open
    const store = config.store ?? createMemorySessionStore();
    return { store };
}

export function createEngineOr(config: EngineConfig) {
    const store = config.store || createMemorySessionStore();
    return { store };
}

export function createEngineUnconditional() {
    // hard-wired, no config hook for a shared store
    const store = createMemorySessionStore();
    return { store };
}

File: harness/semgrep/fixtures/failopen/store.bad.go
package server

// BAD: replay/session store silently defaults to an in-memory implementation
// when the shared/persistent backing is not configured. Multi-instance => fails open.

type Config struct {
	Store SessionStore
}

func NewServer(config Config) *Server {
	if config.Store == nil {
		config.Store = core.NewMemoryStore() // <-- fail-open default (qualified ctor)
	}
	return &Server{store: config.Store}
}

func newSessionMethod(options Options) *Method {
	store := options.Store
	if store == nil {
		store = NewMemoryChannelStore() // <-- fail-open default (short-var form)
	}
	return &Method{store: store}
}
